packages/rolldown/src/plugin/with-filter.ts

```diff
@@ -0,0 +1,42 @@
+import { HookFilterExtension, Plugin } from '..';
+
+type OverrideFilterObject = {
+  transform?: HookFilterExtension<'transform'>['filter'];
+  resolveId?: HookFilterExtension<'resolveId'>['filter'];
+  load?: HookFilterExtension<'load'>['filter'];
+};
+// TODO: maybe we should also change the name as well ?
+// my proposal is `${import.meta.filename}@${name}`
+export function withFilter<A, T extends Plugin<A>>(
+  plugin: T,
+  filter_obj: OverrideFilterObject,
+): T {
```

Comment:
This does not work with nested plugins.
For example, a plugin with a type of `[Plugin, Promise<[Plugin, Promise<Plugin>]>]`.
Also, we need a way to specify which plugin the filter should be set to for those cases.